plugins/lua/buildingplan/planneroverlay.lua

```diff
@@ -452,6 +456,101 @@ local function get_placement_errors()
+-- label: string representation of the filter
+-- mats: list of material names allowed by the filter
+BUILDINGPLAN_FILTERS_KEY = "buildingplan/quick-filters"
+
+QuickFilter = defclass(QuickFilter, widgets.Panel)
+QuickFilter.ATTRS{
+    idx=DEFAULT_NIL,
+    on_click_fn=DEFAULT_NIL,
+    is_selected_fn=function () return false end,
+}
+
+function QuickFilter:init()
+    local label = ('%d.'):format(self.idx)
+    self.frame.w = 27
+    self.renaming = false
+
+    self:addviews {
+        widgets.Label{
+            frame={t=0, l=0},
+            text=string.char(16), -- this is the "►" character
+            text_pen=COLOR_YELLOW,
+            auto_width=true,
+            visible=self.is_selected_fn,
+        },
+        widgets.Label { frame = { t = 0, l = 2 }, text = label },
+        widgets.Label {
+            frame = { t = 0, l = 5 },
+            text = { { text = self:callback('get_label_text'), pen = function() return COLOR_CYAN end } },
+            width = item_filter_chars + 2,
+            visible = function () return self.renaming == false end,
+            on_click = self:callback("on_click"),
+        },
+        widgets.EditField {
+            view_id='edit_field',
+            frame = { t = 0, l = 5 },
+            text = "",
+            width = item_filter_chars + 2,
+            modal = true,
+            visible = function () return self.renaming == true end,
+            on_submit = function (text) self:submit_name(text) end,
+        },
+        widgets.Label {
+            frame = { t = 0, r = 0, w = 3 },
+            text = "[x]",
+            text_pen = COLOR_LIGHTRED,
+            visible = self:callback("slot_used"),
+            on_click = self:callback("clear")
+        }
+    }
+end
+
+function QuickFilter:on_click()
+    if dfhack.internal.getModifiers().shift and self:slot_used() then
+        self.subviews.edit_field.text = self:get_label_text()
+        self.renaming = true
+        self.subviews.edit_field:setFocus(true)
+    else
+        self.on_click_fn(self.idx) -- save/apply filter based on selected ItemLine
+    end
+end
+
+function QuickFilter:slot_used()
+    local quick_filters = dfhack.persistent.getSiteData(BUILDINGPLAN_FILTERS_KEY, {})
+    return quick_filters[self.idx] ~= nil
+end
+
+function QuickFilter:clear()
+    local quick_filters = dfhack.persistent.getSiteData(BUILDINGPLAN_FILTERS_KEY, {})
+    quick_filters[self.idx] = nil
+    dfhack.persistent.saveSiteData(BUILDINGPLAN_FILTERS_KEY, quick_filters)
+end
+
+function QuickFilter:get_label_text()
+    local quick_filters = dfhack.persistent.getSiteData(BUILDINGPLAN_FILTERS_KEY, {})
+    local set = quick_filters[self.idx]
+    if not set then
+        return "empty"
+    else
+        return set.label
+    end
+end
+
+function QuickFilter:submit_name(text)
+    local quick_filters = dfhack.persistent.getSiteData(BUILDINGPLAN_FILTERS_KEY, {})
+    quick_filters[self.idx].label = compress(text, item_filter_chars+2)
```

Comment:
item_filter_chars+2 is written several times. does it deserve its own named constant?

Reply:
If you want, you can change it from 17 to 19 and add -2 to the two places where its used on its own, but I'm not sure that's much of an improvement. For all uses, except the one you marked above, it accounts for the brackets. Above, it accounts for the possible absence of brackets in user-provided filter labels.